src/backend/kvm/vm/builder.rs

```diff
@@ -14,22 +14,45 @@ use primordial::Page;
 use x86_64::structures::paging::page_table::{PageTable, PageTableFlags};
 use x86_64::{align_up, PhysAddr, VirtAddr};
 
-use std::mem::size_of;
-use std::num::NonZeroUsize;
-
-// It is convenient and cheap to provision up to a maximum
-// thread count.
-const DEFAULT_MAX_CPU: usize = 256;
+use std::mem::{align_of, size_of};
+
+/// The first part of the setup area of the VM
+///
+/// - The first page is a zero page.
+/// - The page tables follow the zero page.
+#[repr(C, align(4096))]
+pub struct SetupRegionPre {
+    zero_page: Page,
+    pml4t: PageTable,
+    pml3t_ident: PageTable,
+}
 
-pub struct Config {
-    pub n_syscall_pages: NonZeroUsize,
+/// The setup area of the VM
+///
+/// The address space needs to be laid out in a way that the shim
+/// is expecting. This means that:
+/// - The first page is a zero page.
+/// - The page tables follow the zero page.
+/// - The rest of the `MAX_SETUP_SIZE` bytes is consumed by an
+///   array of [`sallyport::Block`](crate::sallyport::Block)
+/// - The first shared [`sallyport::Block`](crate::sallyport::Block)
+///   contains the [`BootInfo`](crate::backend::kvm::shim::BootInfo) struct at start.
+///   The [`BootInfo`](crate::backend::kvm::shim::BootInfo)
+///   struct communicates other important values to the shim.
+#[repr(C, align(4096))]
+pub struct SetupRegion {
+    pre: SetupRegionPre,
+    syscall_blocks: [Block; N_SYSCALL_BLOCKS],
 }
 
+pub const N_SYSCALL_BLOCKS: usize =
+    (MAX_SETUP_SIZE - size_of::<SetupRegionPre>()) / size_of::<Block>();
+
+pub struct Config;
```

Comment:
I think `Config` can be removed. Unless we want the number of syscall blocks to be configurable.

Reply:
:+1: